src/optimizer/util.cpp

```diff
@@ -23,6 +26,145 @@ namespace peloton {
 namespace optimizer {
 namespace util {
 
+void FillTransitiveTable(
+    std::vector<AnnotatedExpression> predicates,
+    std::map<PredicateInfo, std::vector<expression::AbstractExpression *>> &transitive_table) {
+  for (auto &predicate : predicates) {
+    if (predicate.expr->GetExpressionType() == ExpressionType::COMPARE_EQUAL) {
+      auto expr = predicate.expr;
+
+      PELOTON_ASSERT(expr->GetChildrenSize() == 2);
+
+      auto child1 = expr->GetModifiableChild(0);
+      auto child2 = expr->GetModifiableChild(1);
+
+      if (child1->GetExpressionType() == ExpressionType::VALUE_TUPLE) {
+	auto child_tuple = dynamic_cast<expression::TupleValueExpression *>(child1);
+	PredicateInfo p_info = PredicateInfo(child_tuple->GetTableName(), child_tuple->GetColumnName());
+
+	bool not_found = true;
+	auto predicate_to_add = expr->GetModifiableChild(1);
+	auto predicate_list = transitive_table[p_info];
+	for (auto predicate : predicate_list) {
+	  if (predicate->ExactlyEquals(*predicate_to_add)) {
+	    not_found = false;
+	    break;
+	  }
+	}
+
+	if (not_found) {
+	  transitive_table[p_info].push_back(predicate_to_add);
+	}
+      }
+
+      if (child2->GetExpressionType() == ExpressionType::VALUE_TUPLE) {
+	auto child_tuple = dynamic_cast<expression::TupleValueExpression *>(child2);
+	PredicateInfo p_info = PredicateInfo(child_tuple->GetTableName(), child_tuple->GetColumnName());
+
+	bool not_found = true;
+	auto predicate_to_add = expr->GetModifiableChild(0);
+	auto predicate_list = transitive_table[p_info];
+	for (auto predicate : predicate_list) {
+	  if (predicate->ExactlyEquals(*predicate_to_add)) {
+	    not_found = false;
+	    break;
+	  }
+	}
+
+	if (not_found) {
+	  transitive_table[p_info].push_back(predicate_to_add);
+	}
+      }
+    }
+  }
+}
+
+std::vector<AnnotatedExpression> GenerateTransitivePredicates(
+    std::vector<AnnotatedExpression> predicates,
+    std::map<PredicateInfo, std::vector<expression::AbstractExpression *>> &transitive_table) {
+  std::vector<AnnotatedExpression> new_predicates;
+
+  for (auto predicate : predicates) {
+    std::string exp_type = ExpressionTypeToString(predicate.expr->GetExpressionType());
+    if (exp_type.substr(0, 7) == "COMPARE") {
```

Comment:
i'd try and avoid using magic number (`7` in this case)

Reply:
Agreed, you could probably do it like how it is done [here](https://github.com/cmu-db/peloton/pull/1340/files#diff-d1982d39ac0cf4b4426e20aa7f9c0b6aR225)